README.md
[![Build Status](https://travis-ci.org/Chrezm/TsuserverDR.svg?branch=master)](https://travis-ci.org/Chrezm/TsuserverDR)
# TsuserverDR

A Python-based server for Danganronpa Online. It is a fork from [tsuserver3](https://github.com/AttorneyOnline/tsuserver3) which is targeted towards Attorney Online.

Requires Python 3.7-3.10 and PyYAML (follow instructions below to install).

## How to use

### Installing

It is highly recommended you read through all the installation steps first before going through them.

* Install the latest version of Python. **Python 2 will not work**.
  - You can download Python from its official website [here](https://www.python.org/downloads/). If you already have Python installed, check that your Python version satisfies the version requirement listed above.
  - If prompted during the installation to add `python` as a PATH environment variable, accept this option. You may see this option appear on the very first screen during the installation process.
  - If you know what a virtual environment is and your system supports it, it is recommended that you use one, such as [Anaconda](https://www.continuum.io/downloads) for Windows, or [virtualenv](https://virtualenv.pypa.io/en/stable/) for everyone else (it runs itself using Python). If you do not know what a virtual environment is, you may skip this point.
  - If you have Python 3.6 or lower, you will be prompted on server launch to update to a newer version of Python. That is because the server requires Python 3.7 or higher. Follow instructions under Updating to update your Python version.

* Open Command Prompt, PowerShell or your preferred terminal, and change to the directory where you downloaded TsuserverDR to. You can do this in two ways:
  - On Windows 10, go up one folder above the TsuserverDR folder, Shift + right click the TsuserverDR folder, and click `Open PowerShell window here`. This is the easiest method. Alternatively...
  - On most operating systems, copy the path of the TsuserverDR folder, open the terminal, and type in `cd "[paste here]"`, excluding the brackets, but including the quotation marks if the path contains spaces.

* Install PyYAML and dependencies by typing in the following two commands in the terminal you just opened:
  ```
  python -m pip install --upgrade pip
  python -m pip install --user -r requirements.txt
  ```

  If you are using Windows and have both Python 2 and 3 installed, you may do the following:
  ```
  py -3 -m pip install --upgrade pip
  py -3 -m pip install --user -r requirements.txt
  ```

  This operation should not require administrator privileges, unless you decide to remove the `--user` option.
* Rename the folder `config_sample` to `config` and edit the values in the provided YAML files to your liking. Be sure to check the YAML files for syntax errors after you are done. *Use spaces only; do not use tabs.*

### Running

* To launch a server, you may either
  - Double-click `start_server.py` in your TsuserverDR folder, or...
  - In PowerShell, Command Prompt or your preferred terminal, change directory to your TsuserverDR folder and type `python start_server.py`, or `py -3 start_server.py` if you use both Python 2 and 3. For instructions on how to launch any of the above programs or change directory, refer to the second point in the Installing section.

* If everything was set up correctly, you will see something like this appear:

```
[2022-05-05T10:20:20]: Starting...
[2022-05-05T10:20:20]: Launching TsuserverDR 4.3.1 (220505a)...
[2022-05-05T10:20:20]: Loading server configurations...
[2022-05-05T10:20:20]: Server configurations loaded successfully!
[2022-05-05T10:20:20]: Starting a nonlocal server...
[2022-05-05T10:20:20]: Server started successfully!
[2022-05-05T10:20:21]: Server should be now accessible from 192.0.2.0:50000:My First DR Server
```

* If you are listing your server in the Attorney Online master server, make sure its details are set up correctly. In particular, make sure that your server name and description are correct, as that is how players will find your server. If everything was set up correctly, you will see something like this appear:

```
[2022-05-05T10:20:21]: Attempting to connect to the master server at master.aceattorneyonline.com:27016 with the following details:
[2022-05-05T10:20:21]: *Server name: My First DR Server
[2022-05-05T10:20:21]: *Server description: This is my flashy new DR server
[2022-05-05T10:20:22]: Connected to the master server.
```

  - The server will make a single ping to [ipify](https://api.ipify.org) in order to obtain its public IP address. If it fails to do that, it will let you know that, as it means there is probably something wrong with your internet connection and that other players may not be able to connect to your server.
  - Successful connection or getting a spot in the master server list does not imply that your server will be accessible to other players. In particular, you must make sure that your external port in `config\config.yaml` is open and accepting connections, which usually involves a combination of router and firewall settings. In case of doubt, you can use websites such as [Can You See Me](https://canyouseeme.org) to check if your port is visible.

* To stop the server, press Ctrl+C once from your terminal. This will initiate a shutdown sequence and notify you when it is done. If the shutdown finished successfully, you will see something like this appear:

```
[2022-05-05T22:23:04]: You have initiated a server shut down.
[2022-05-05T22:23:04]: Kicking 12 remaining clients.
[2022-05-05T22:23:04]: Server has successfully shut down.
```

* If you do not see anything after a few seconds of starting a shutdown, you can try spamming Ctrl+C to try and force a shutdown or directly close out your terminal. This is not recommended due to the cleanup process not finishing correctly but it is doable.

* To restart a server, you can follow the steps outlined above to shut down and then start the server again.

* In the unlikely event that there is an error during runtime, the server will do its best to print out to your terminal a complete traceback of the error with some additional information, as well as a more complete log file in the `logs` folder with the error timestamp, to help with debugging. Depending on the nature of the error, the server may or may not be able to continue execution normally after such an error happens.

### Updating

* If you already have a version of TsuserverDR or tsuserver3 installed and wish to update to a new version, you can download the new version and then overwrite your previously existing files. Do note that you do not need to shut down your server before overwriting the files, but you must restart it from console in order for changes to take effect.

  - This process will not overwrite your server configurations inside the `config` folder, your existing logs inside the `logs` folder, or the user information inside the `storage` folder. However, it will overwrite other files including the Python files inside the `server` folder. Therefore, make sure to save backups of those files before overwriting in case you have modified them and wish to keep an archive of your changes.

* If you want to update **Python** itself, you can get the latest Python download [from their website here](https://www.python.org/downloads/) and then follow the instructions under the Installing section in this readme. To check your current version of Python, you can run ``python`` on its own and look at the first line. The latest stable major Python release is *Python 3.9* as of October 5, 2020.

  - Please follow the installation instructions again even if you had successfully ran a server before, because your new Python installation may be missing libraries that TsuserverDR expects there to exist. You should not need to change any server configuration files though.

CHANGELOG.md
## TsuserverDR 3
### 180801 (3.0.1)
* Added /time12

### 180812 (3.1)
* Added /ddroll, /ddrollp for rolls with modifiers
* Added 'rp_getarea(s)' attributes to restrict /getarea(s)
* Added /toggle_rpgetarea(s) to change the above attribute
* Added /st for staff-only chat
* /g now includes OOC usernames as opposed to charname

### 180910 (3.2)
* Added reachable areas support with
  - /unilock
  - /bilock
  - /delete_areareachlock
  - /restore_areareachlock
* Added proper support for spectators (can talk OOC/use commands/change music&areas/use judge buttons now)
* Added support for IC locking with /iclock
* OOC usernames may no longer include the hostname or global message prefixe
* Fixed /kick crash on unrecognized parameters

### 190126 (3.3)
* Added AFK kicks
* Added user-initiated timers with
  - /timer start
  - /timer get
  - /timer cancel
* Area lists now only list reachable areas

### 190225 (3.4)
* Added inter-area OOC communication with /scream
* Added debugging tool /exec
* Added inter-area music setting with /rplay
* Renamed /ddroll -> /roll, /ddrollp -> /rollp

### 190323 (3.5)
* Added sneak support with /sneak and /reveal
* Added inter-area IC announcements with /globalic, /unglobalic
* Added server custom shownames with /shownames
* Added soundproof area attribute
* Expanded inter-area OOC communication with /knock

### 190519 (3.6)
* Added support for custom music lists with /music_list
* Added area list reloading with /area_list
* Added character restriction on a per-area basis with /char_restrict
* Added scream ranges on a per-area basis (deprecating soundproof attribute)
* Default server configuration now uses DRO standards

### 190602 (3.7)
* Added automatic passing messages with /autopass
* Added passage lock transientness with /transient
* Added showname moderation tools
  - /showname_freeze
  - /showname_history
  - /showname_list
  - /showname_nuke
  - /showname_set
* Enforced showname uniqueness per area
* Fixed /area_kick crash on unrecognized parameters
* Fixed /getareas listing an empty area if only people there are sneaking
* Players on server select no longer count in online playercount, nor in /getarea[s] and /area

### 190609 (3.8)
* Added movement handicaps with /handicap, /unhandicap
* Added area lights with /lights
* Added blood trails with
  - /bloodtrail
  - /bloodtrail_clean
  - /bloodtrail_list
  - /bloodtrail_set
* Added /toggle_shownames to disable receiving server shownames
* Allowed for more DRO client effects being used
* Fixed /autopass not sending messages from or for staff

### 190618 (first with date version system) (3.9)
* Added area descriptions with /look, /look_set
* Console now displays server opening/closing/reconnecting messages.
* Python errors now show additional debugging info on server log+client if error came as a result of command
* Renamed some commands
  - allow_iniswap -> /can_iniswap
  - delete_areareachlock -> /passage_clear
  - restore_areareachlock -> /passage_restore
  - toggle_areareachlock -> /can_passagelock
  - toggleglobal -> /toggle_global
  - toggle_rollp -> /can_rollp
  - toggle_rpgetarea -> /can_rpgetarea
  - toggle_rpgetareas -> /can_rpgetareas
* Reworked /area_kick so that it now allows area names as parameter
* Reworked /discord so that it now shows the server's Discord server invite link
* Fixed /autopass sending old character name in case of forced char switch on area change
* Fixed /bloodtrail_set not enforcing staff member rank
* Fixed /disemconsonant and /remove_h not working/crashing
* Fixed /follow allowing to follow yourself
* Fixed "Knock" client effect not being sent

### 190630 (3.10)
* Added support for user initiated day cycles with
  - /clock
  - /clock_cancel
  - /clock_pause
  - /clock_unpause
* Added various moderation tools
  - /judgelog
  - /shoutlog
  - /version
  - /whereis
* Added 'has_lights' parameters to areas to allow/disallow using /lights

server/clients.py
# TsuserverDR, a Danganronpa Online server based on tsuserver3, an Attorney Online server
#
# Copyright (C) 2016 argoneus <argoneuscze@gmail.com> (original tsuserver3)
# Current project leader: 2018-22 Chrezm/Iuvee <thechrezm@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

from server.constants import ArgType

class _Singleton():
    def __new__(cls):
        if not hasattr(cls, 'instance'):
            cls.instance = super(_Singleton, cls).__new__(cls)
        return cls.instance

class DefaultDROProtocol(_Singleton):
    def __eq__(self, other):
        return type(self).__name__ == type(other).__name__

    HAS_CLIENTSIDE_MUSIC_LOOPING = True
    HAS_DISTINCT_AREA_AND_MUSIC_LIST_OUTGOING_PACKETS = True

    DECRYPTOR_OUTBOUND = [
        ('key', 34),  # 0
        ]

    HI_INBOUND = [
        ('client_hdid', ArgType.STR),  # 0
        ]

    BD_OUTBOUND = [
        ]

    ID_OUTBOUND = [
        ('client_id', 0),  # 0
        ('server_software', 'TsuserverDR'),  # 1
        ('server_software_version', '0.0.0'),  # 2
        ]

    PN_OUTBOUND = [
        ('player_count', 0),  # 0
        ('player_limit', 0),  # 1
        ]

    ID_INBOUND = [
        ('client_software', ArgType.STR),  # 0
        ('client_software_version', ArgType.STR),  # 1
        ]

    FL_OUTBOUND = [
        ('fl_ao2_list', list()),  # 0
        ]

    CLIENT_VERSION_OUTBOUND = [
        ('dro_version_ao2_list', list()),  # 0
    ]

    CH_INBOUND = [
        ('char_id', ArgType.INT),  # 0
        ]

    CHECK_OUTBOUND = [
        ]

    ASKCHAA_INBOUND = [
        ]

    SI_OUTBOUND = [
        ('char_count', 0),  # 0
        ('evidence_count', 0),  # 1
        ('music_list_count', 0),  # 2
        ]

    AE_INBOUND = [
        ('evidence_page', ArgType.INT),  # 0
        ]

    RC_INBOUND = [
        ]

    SC_OUTBOUND = [
        ('chars_ao2_list', list()),  # 0
        ]

    RM_INBOUND = [
        ]

    SM_OUTBOUND = [
        ('music_ao2_list', list()),  # 0
        ]

    RD_INBOUND = [
        ]

    CHARSCHECK_OUTBOUND = [
        ('chars_status_ao2_list', list()),  # 0
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ('tod_backgrounds_ao2_list', list()),  # 1
        ]

    LE_OUTBOUND = [
        ('evidence_ao2_list', list()),  # 0
        ]

    MM_OUTBOUND = [
        ('unknown', 1),  # 0
        ]

    OPPASS_OUTBOUND = [
        ('guard_pass', ''),  # 0
        ]

    DONE_OUTBOUND = [
        ]

    FM_OUTBOUND = [
        ('music_ao2_list', list()),  # 0
        ]

    FA_OUTBOUND = [
        ('areas_ao2_list', list()),  # 0
        ]

    CC_INBOUND = [
        ('client_id', ArgType.INT),
        ('char_id', ArgType.INT),
        ('client_hdid', ArgType.STR),
        ]

    PV_OUTBOUND = [
        ('client_id', 0),  # 0
        ('char_id_tag', 'CID'),  # 1
        ('char_id', -1),  # 2
        ]

    CT_INBOUND = [
        ('username', ArgType.STR),  # 0
        ('message', ArgType.STR),  # 1
        ]

    CT_OUTBOUND = [
        ('username', ''),  # 0
        ('message', ''),  # 1
        ]

    ACKMS_OUTBOUND = [
        ]

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR_OR_EMPTY),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ('showname', ArgType.STR_OR_EMPTY),  # 15
        ('video', ArgType.STR_OR_EMPTY),  # 16
        ('hide_character', ArgType.INT),  # 17
        ]

    MS_OUTBOUND = [
        ('msg_type', 0),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', -1),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ''),  # 15
        ('video', '0'),  # 16
        ('hide_character', 0),  # 17
        ('client_id', -1),  # 18
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ('force_same_restart', 1),  # 3
        ]

    RT_INBOUND = [
        ('name', ArgType.STR),  # 0
        ]

    RT_OUTBOUND = [
        ('name', ''),  # 0
        ]

    HP_INBOUND = [
        ('side', ArgType.INT),  # 0
        ('health', ArgType.INT),  # 1
        ]

    HP_OUTBOUND = [
        ('side', 1),  # 0
        ('health', 0),  # 1
        ]

    PE_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('description', ArgType.STR),  # 1
        ('image', ArgType.STR),  # 2
        ]

    DE_INBOUND = [
        ('evi_id', ArgType.INT),  # 0
        ]

    EE_INBOUND = [
        ('evi_id', ArgType.INT),  # 0
        ('name', ArgType.STR),  # 1
        ('description', ArgType.STR),  # 2
        ('image', ArgType.STR),  # 3
        ]

    ZZ_INBOUND = [
        ]

    ZZ_OUTBOUND = [
        ('message', ''),  # 0
        ]

    SP_OUTBOUND = [
        ('position', ''),  # 0
        ]

    CL_OUTBOUND = [
        ('client_id', 0),  # 0
        ('hour', 0),  # 0
        ]

    TR_OUTBOUND = [
        ('timer_id', 0),  # 0
        ]

    TP_OUTBOUND = [
        ('timer_id', 0),  # 0
        ]

    TST_OUTBOUND = [
        ('timer_id', 0),  # 0
        ('new_time', 0),  # 1
        ]

    TSS_OUTBOUND = [
        ('timer_id', 0),  # 0
        ('new_step_length', 0),  # 1
        ]

    TSF_OUTBOUND = [
        ('timer_id', 0),  # 0
        ('new_firing_interval', 0),  # 1
        ]

    GM_OUTBOUND = [
        ('name', ''),  # 0
        ]

    TOD_OUTBOUND = [
        ('name', ''),  # 0
        ]

    SN_INBOUND = [
        ('showname', ArgType.STR_OR_EMPTY),  # 0
        ]

    SN_OUTBOUND = [
        ('showname', ''),  # 0
        ]

    CHRINI_INBOUND = [
        ('actual_folder_name', ArgType.STR),  # 0
        ('actual_character_showname', ArgType.STR),  # 1
        ]

    CHAT_TICK_RATE_OUTBOUND = [
        ('chat_tick_rate', -1),  # 0
        ]

    CHARSCHECK_INBOUND = [
        ]

class ClientDRO1d1d0(DefaultDROProtocol):
    pass

class ClientDRO1d0d0(DefaultDROProtocol):
    HAS_CLIENTSIDE_MUSIC_LOOPING = False
    HAS_DISTINCT_AREA_AND_MUSIC_LIST_OUTGOING_PACKETS = False

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR_OR_EMPTY),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
    ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', -1),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ' '),  # 15
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ]


class ClientDROLegacy(DefaultDROProtocol):
    HAS_CLIENTSIDE_MUSIC_LOOPING = False
    HAS_DISTINCT_AREA_AND_MUSIC_LIST_OUTGOING_PACKETS = False

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', 0),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ' '),  # 15
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ]


class ClientAO2d6(DefaultDROProtocol):
    HAS_CLIENTSIDE_MUSIC_LOOPING = False
    HAS_DISTINCT_AREA_AND_MUSIC_LIST_OUTGOING_PACKETS = False

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ('showname', ArgType.STR_OR_EMPTY),  # 15
        ('charid_pair', ArgType.INT),  # 16
        ('offset_pair', ArgType.INT),  # 17
        ('nonint_pre', ArgType.INT),  # 18
        ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', 0),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ' '),  # 15
        ('charid_pair', -1),  # 16
        ('other_folder', ''),  # 17
        ('other_emote', ''),  # 18
        ('offset_pair', 0),  # 19
        ('other_offset', 0),  # 20
        ('other_flip', 0),  # 21
        ('nonint_pre', 0),  # 22
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ('showname', ArgType.STR_OR_EMPTY),  # 2
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ]


class ClientAO2d7(DefaultDROProtocol):
    HAS_CLIENTSIDE_MUSIC_LOOPING = False
    HAS_DISTINCT_AREA_AND_MUSIC_LIST_OUTGOING_PACKETS = False

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ('showname', ArgType.STR_OR_EMPTY),  # 15
        ('charid_pair', ArgType.INT),  # 16
        ('offset_pair', ArgType.INT),  # 17
        ('nonint_pre', ArgType.INT),  # 18
        ('looping_sfx', ArgType.INT),  # 19
        ('screenshake', ArgType.INT),  # 20
        ('frame_screenshake', ArgType.STR_OR_EMPTY),  # 21
        ('frame_realization', ArgType.STR_OR_EMPTY),  # 22
        ('frame_sfx', ArgType.STR_OR_EMPTY),  # 23
        ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', 0),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ''),  # 15
        ('charid_pair', -1),  # 16
        ('other_folder', ''),  # 17
        ('other_emote', ''),  # 18
        ('offset_pair', 0),  # 19
        ('other_offset', 0),  # 20
        ('other_flip', 0),  # 21
        ('nonint_pre', 0),  # 22
        ('looping_sfx', 0),  # 23
        ('screenshake', 0),  # 24
        ('frame_screenshake', ''),  # 25
        ('frame_realization', ''),  # 26
        ('frame_sfx', ''),  # 27
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ('showname', ArgType.STR_OR_EMPTY),  # 2
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ]

    FA_OUTBOUND = [
        ('areas_ao2_list', list()),  # 0
        ]


class ClientAO2d8d4(DefaultDROProtocol):
    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ('showname', ArgType.STR_OR_EMPTY),  # 15
        ('charid_pair_pair_order', ArgType.STR),  # 16
        ('offset_pair', ArgType.INT),  # 17
        ('nonint_pre', ArgType.INT),  # 18
        ('looping_sfx', ArgType.INT),  # 19
        ('screenshake', ArgType.INT),  # 20
        ('frame_screenshake', ArgType.STR_OR_EMPTY),  # 21
        ('frame_realization', ArgType.STR_OR_EMPTY),  # 22
        ('frame_sfx', ArgType.STR_OR_EMPTY),  # 23
        ('additive', ArgType.INT),  # 24
        ('effect', ArgType.STR),  # 25
        ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', 0),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ''),  # 15
        ('charid_pair_pair_order', -1),  # 16
        ('other_folder', ''),  # 17
        ('other_emote', ''),  # 18
        ('offset_pair', 0),  # 19
        ('other_offset', 0),  # 20
        ('other_flip', 0),  # 21
        ('nonint_pre', 0),  # 22
        ('looping_sfx', 0),  # 23
        ('screenshake', 0),  # 24
        ('frame_screenshake', ''),  # 25
        ('frame_realization', ''),  # 26
        ('frame_sfx', ''),  # 27
        ('additive', 0),  # 28
        ('effect', ''),  # 29
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ('showname', ArgType.STR_OR_EMPTY),  # 2
        ('effects', ArgType.INT),  # 3
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ('loop', 1),  # 3
        ('channel', 0),  # 4
        ('effects', 0),  # 5
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ('pos', ''),  # 1
        ]


class ClientAO2d9d0(DefaultDROProtocol):
    ASKCHAA_INBOUND = [
        ('ao290doesnotsupportpacketswithnoarguments', ArgType.STR_OR_EMPTY),  # 0
        ]

    RC_INBOUND = [
        ('ao290doesnotsupportpacketswithnoarguments', ArgType.STR_OR_EMPTY),  # 0
        ]

    RM_INBOUND = [
        ('ao290doesnotsupportpacketswithnoarguments', ArgType.STR_OR_EMPTY),  # 0
        ]

    RD_INBOUND = [
        ('ao290doesnotsupportpacketswithnoarguments', ArgType.STR_OR_EMPTY),  # 0
        ]

    ZZ_INBOUND = [
        ('ao290doesnotsupportpacketswithnoarguments', ArgType.STR_OR_EMPTY),  # 0
        ]

    MS_INBOUND = [
        ('msg_type', ArgType.STR),  # 0
        ('pre', ArgType.STR_OR_EMPTY),  # 1
        ('folder', ArgType.STR),  # 2
        ('anim', ArgType.STR),  # 3
        ('text', ArgType.STR),  # 4
        ('pos', ArgType.STR),  # 5
        ('sfx', ArgType.STR),  # 6
        ('anim_type', ArgType.INT),  # 7
        ('char_id', ArgType.INT),  # 8
        ('sfx_delay', ArgType.INT),  # 9
        ('button', ArgType.INT),  # 10
        ('evidence', ArgType.INT),  # 11
        ('flip', ArgType.INT),  # 12
        ('ding', ArgType.INT),  # 13
        ('color', ArgType.INT),  # 14
        ('showname', ArgType.STR_OR_EMPTY),  # 15
        ('charid_pair_pair_order', ArgType.STR),  # 16
        ('offset_pair', ArgType.STR),  # 17
        ('nonint_pre', ArgType.INT),  # 18
        ('looping_sfx', ArgType.INT),  # 19
        ('screenshake', ArgType.INT),  # 20
        ('frame_screenshake', ArgType.STR_OR_EMPTY),  # 21
        ('frame_realization', ArgType.STR_OR_EMPTY),  # 22
        ('frame_sfx', ArgType.STR_OR_EMPTY),  # 23
        ('additive', ArgType.INT),  # 24
        ('effect', ArgType.STR),  # 25
        ]

    MS_OUTBOUND = [
        ('msg_type', 1),  # 0
        ('pre', '-'),  # 1
        ('folder', '<NOCHAR>'),  # 2
        ('anim', '../../misc/blank'),  # 3
        ('msg', ''),  # 4
        ('pos', 'jud'),  # 5
        ('sfx', 0),  # 6
        ('anim_type', 0),  # 7
        ('char_id', 0),  # 8
        ('sfx_delay', 0),  # 9
        ('button', 0),  # 10
        ('evidence', 0),  # 11
        ('flip', 0),  # 12
        ('ding', -1),  # 13
        ('color', 0),  # 14
        ('showname', ''),  # 15
        ('charid_pair_pair_order', -1),  # 16
        ('other_folder', ''),  # 17
        ('other_emote', ''),  # 18
        ('offset_pair', 0),  # 19
        ('other_offset', 0),  # 20
        ('other_flip', 0),  # 21
        ('nonint_pre', 0),  # 22
        ('looping_sfx', 0),  # 23
        ('screenshake', 0),  # 24
        ('frame_screenshake', ''),  # 25
        ('frame_realization', ''),  # 26
        ('frame_sfx', ''),  # 27
        ('additive', 0),  # 28
        ('effect', ''),  # 29
        ]

    MC_INBOUND = [
        ('name', ArgType.STR),  # 0
        ('char_id', ArgType.INT),  # 1
        ('showname', ArgType.STR_OR_EMPTY),  # 2
        ('effects', ArgType.INT),  # 3
        ]

    MC_OUTBOUND = [
        ('name', ''),  # 0
        ('char_id', -1),  # 1
        ('showname', ''),  # 2
        ('loop', 1),  # 3
        ('channel', 0),  # 4
        ('effects', 0),  # 5
        ]

    BN_OUTBOUND = [
        ('name', ''),  # 0
        ('pos', ''),  # 1
        ]

class ClientKFO2d8(ClientAO2d7):
    pass

class ClientCC22(ClientAO2d6):
    pass

class ClientCC24(ClientAO2d8d4):
    pass
